src/classes/RD2_ValidationService_TEST.cls

```diff
@@ -0,0 +1,198 @@
+            .withAccount(acc.Id)
+            .build();
+        try {
+            insert rd;
+        } catch (Exception e) {
+            errMessage = e.getMessage();
+        }
+
+        System.assert(errMessage.contains(System.Label.RD_ContactMustBelongToAccount));
+
+    }
+
+    /***
+    * @description Verifies if it displays error message when the contact do not belong to
+        the selected HouseHold account.
+    */
+    @isTest
+    private static void shouldFailWhenContactDoesNotBelongToHHAccount() {
+        String errMessage = '';
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        Account acc = getHHAccount();
+        insert acc;
+
+        Contact con = UTIL_UnitTestData_TEST.getContact();
+        insert con;
+
+        npe03__Recurring_Donation__c rd = new TEST_RecurringDonationBuilder()
+            .withAccount(acc.Id)
+            .withContact(con.Id)
+            .build();
+        try {
+            insert rd;
+        } catch (Exception e) {
+            errMessage = e.getMessage();
+        }
+        System.assert(errMessage.contains(System.Label.RD_ContactMustBelongToAccount));
+    }
+
+    /**
+    * @description Should fail when there is an invalid contact for the account in the
+        Recurring Donations list.
+    */
+    @isTest
+    private static void shouldFailOnlyInvalidRecurringDonations() {
+
+        String errMessage ='';
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        List<Contact> contacts = new List<Contact>{
+            UTIL_UnitTestData_TEST.getContact(),
+            UTIL_UnitTestData_TEST.getContact()
+        };
+        insert contacts;
+
+        contacts = [SELECT AccountId FROM Contact WHERE Id IN :contacts];
+
+        List<npe03__Recurring_Donation__c> rds = new TEST_RecurringDonationBuilder()
+        .withAccount(contacts[0].AccountId)//only RD for one Contact will be valid
+        .withAmount(100)
+        .build(contacts);
+
+        List<Database.SaveResult> results = Database.insert(rds, false);
+        ERR_Handler.Errors errors = ERR_Handler.getErrors(results, rds);
+        System.assertEquals(true, errors.errorsExist);
+        System.assertEquals(1, errors.errorRecords.size());
+        System.assert([
+            SELECT id
+            FROM npe03__Recurring_Donation__c].size()>0 );
+
+    }
+
+    /***
+    * @description Verifies if it creates Recurring donations for a single household contact
+        without an account.
+    */
+    @isTest
+    private static void shouldCreateRDSForHouseholdContact() {
+        RD2_EnablementService_TEST.setRecurringDonations2Enabled();
+
+        Contact con = UTIL_UnitTestData_TEST.getContact();
+        insert con;
+
+        npe03__Recurring_Donation__c rd = new TEST_RecurringDonationBuilder()
+            .withContact(con.Id)
+            .build();
+
+        insert rd;
+
+        System.assert([
```

Comment:
Instead of the assert on the record, we can `insert rd` in the `try/catch` block and verify `System.assert(String.isBlank(errMessage), 'Recurring Donation should be created: ' + errMessage);`.

The same is applicable to `shouldCreateRDSForOrganizationAccounts()`.

Reply:
@RadSFDeveloper , You suggest not to check if the record have been created or not, here we are checking if no error messages only.

Reply:
We would need to check that no exception is thrown when the RD is created:
```
String errMessage = '';
try {
    insert rd;
} catch (Exception e) {
    errMessage = e.getMessage();
}
System.assert(String.isBlank(errMessage), 'Recurring Donation should be created successfully: ' + errMessage);